scripts/builtin/xgboost.dml

```diff
@@ -0,0 +1,780 @@
+    {
+      best_gain = current_gain
+      best_split = current_split
+      best_sim_score_left = sim_score_left
+      best_sim_score_right = sim_score_right
+    }
+  }
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    X: a 1xn matrix (one feature with all values)
+#           value: the border to separate the values
+#           loop: boolean to check if used in loop and returns then other values
+# OUTPUT:   left: a 1xn matrix with all values smaller than "value"
+#           right: a 1xn matrix with all values larger than "value"
+splitMatrixByValue = function(Matrix[Double] X, Double value) return (Matrix[Double] left, Matrix[Double] right) {
+  assert(ncol(X) == 1)
+  left = matrix(0, rows=0, cols=1)
+  right = matrix(0, rows=0, cols=1)
+
+  for(i in 1:nrow(X)) {
+    if(as.scalar(X[i,]) < value)
+    {
+      left = rbind(left,X[i,])
+    }
+    else
+    {
+      right = rbind(right, X[i,])
+    }
+  }
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    curr_X: a 1xn matrix (one feature with current values)
+#           X:     a 1xn matrix the feature of the X-Matrix with all values
+#           value: the border to separate the values
+# OUTPUT:   left: a 1xn matrix with 1.0 if the value is smaller or 0.0 if it is bigger
+#           right: a 1xn matrix with 1.0 if the value is bigger or 0.0 if it is smaller
+splitMatrixByValueLoop = function(Matrix[Double] curr_X, Matrix[Double] X, Double value) return (Matrix[Double] left, Matrix[Double] right) {
+  left = matrix(0, rows=0, cols=1)
+  right = matrix(0, rows=0, cols=1)
+  one_vec = matrix(1, rows=1, cols=1)
+  zero_vec = matrix(0, rows=1, cols=1)
+
+  for(i in 1:nrow(X)) {
+    if (as.scalar(X[i,]) == as.scalar(curr_X[i,]))  # if same row in curr_X and X
+    {
+      if (as.scalar(X[i,]) < value)
+      {
+        left = rbind(left, one_vec[1,]) # add 1 to left and 0 to right vector
+        right = rbind(right, zero_vec[1,])
+      }
+      else
+      {
+        right = rbind(right, one_vec[1,]) # add 1 to right and 0 to left vector
+        left = rbind(left, zero_vec[1,])
+      }
+    }
+    else
+    {
+      left = rbind(left, zero_vec[1,]) # add 0 to both vectors
+      right = rbind(right, zero_vec[1,])
+    }
+  }
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    curr_X: a 1xn matrix (one feature with current values)
+#           X:     a 1xn matrix the feature of the X-Matrix with all values
+# OUTPUT:   left: a 1xn matrix with 1.0 for if the value is true or 0.0 if it is false
+#           right: a 1xn matrix with 1.0 for if the value is false or 0.0 if it is true
+splitMatrixByCategory = function(Matrix[Double] curr_X, Matrix[Double] X) return (Matrix[Double] left, Matrix[Double] right) {
```

Comment:
Also here, it is expensive, 
in this instance simply allocate the output matrices, filled with 0
and then use vector processing:

i think the correct way is:

left = curr_X == 1
right = left != 1

Reply:
improved in https://github.com/apache/systemds/pull/1334/commits/d9d46d0080ce8dd93305205e9977d3448c78b067